src/Control/Effect/Random.hs

```diff
@@ -10,44 +10,88 @@ module Control.Effect.Random
 , MonadInterleave(..)
 ) where
 
+import Control.Applicative (Alternative(..))
 import Control.Effect.Carrier
-import Control.Effect.Internal
-import Control.Effect.Random.Internal
+import Control.Effect.State
 import Control.Effect.Sum
+import Control.Monad (MonadPlus(..))
+import Control.Monad.Fail
 import Control.Monad.Random.Class (MonadInterleave(..), MonadRandom(..))
 import Control.Monad.IO.Class (MonadIO(..))
+import Control.Monad.Trans.Class
 import qualified System.Random as R (Random(..), RandomGen(..), StdGen, newStdGen)
 
+data Random m k
+  = forall a . R.Random a => Random (a -> k)
+  | forall a . R.Random a => RandomR (a, a) (a -> k)
+  | forall a . Interleave (m a) (a -> k)
+
+deriving instance Functor (Random m)
+
+instance HFunctor Random where
+  hmap _ (Random k) = Random k
+  hmap _ (RandomR r k) = RandomR r k
+  hmap f (Interleave m k) = Interleave (f m) k
+  {-# INLINE hmap #-}
+
+instance Effect Random where
+  handle state handler (Random    k) = Random    (handler . (<$ state) . k)
+  handle state handler (RandomR r k) = RandomR r (handler . (<$ state) . k)
+  handle state handler (Interleave m k) = Interleave (handler (m <$ state)) (handler . fmap k)
+
+
 -- | Run a random computation starting from a given generator.
 --
 --   prop> run (runRandom (PureGen a) (pure b)) == (PureGen a, b)
-runRandom :: (Carrier sig m, Effect sig, Monad m, R.RandomGen g) => g -> Eff (RandomC g m) a -> m (g, a)
-runRandom g = flip runRandomC g . interpret
+runRandom :: g -> RandomC g m a -> m (g, a)
+runRandom g = runState g . runRandomC
 
 -- | Run a random computation starting from a given generator and discarding the final generator.
 --
 --   prop> run (evalRandom (PureGen a) (pure b)) == b
-evalRandom :: (Carrier sig m, Effect sig, Monad m, R.RandomGen g) => g -> Eff (RandomC g m) a -> m a
+evalRandom :: Functor m => g -> RandomC g m a -> m a
 evalRandom g = fmap snd . runRandom g
 
 -- | Run a random computation starting from a given generator and discarding the final result.
 --
 --   prop> run (execRandom (PureGen a) (pure b)) == PureGen a
-execRandom :: (Carrier sig m, Effect sig, Monad m, R.RandomGen g) => g -> Eff (RandomC g m) a -> m g
+execRandom :: Functor m => g -> RandomC g m a -> m g
 execRandom g = fmap fst . runRandom g
 
 -- | Run a random computation in 'IO', splitting the global standard generator to get a new one for the computation.
-evalRandomIO :: (Carrier sig m, Effect sig, MonadIO m) => Eff (RandomC R.StdGen m) a -> m a
+evalRandomIO :: MonadIO m => RandomC R.StdGen m a -> m a
 evalRandomIO m = liftIO R.newStdGen >>= flip evalRandom m
 
-newtype RandomC g m a = RandomC { runRandomC :: g -> m (g, a) }
+newtype RandomC g m a = RandomC { runRandomC :: StateC g m a }
```

Comment:
Per #115, I’d like to extract `Random`/`RandomC` from the lib altogether, but that’s fodder for another PR. In the meantime, we are not currently providing `MonadRandom`/`MonadInterleave` interfaces for any of the other carriers in the lib.